src/junglescout/client_base.py

```diff
@@ -0,0 +1,52 @@
+from typing import Optional, NoReturn
+import httpx
+
+from junglescout.models.parameters import ApiType, Marketplace
+import urllib.parse
+
+
+class BaseClient:
```

Comment:
Consider making `BaseClient` an abstract base class. Also, consider introducing `session` to the interface as a generic, since every client will need a `session` instance.

Reply:
I have something like this in mind

```
@property
@abstractmethod
def session(self):
  return self._session
```

Is that what you have in mind?

Reply:
I was thinking something like this for the `BaseClient`:

```python
from typing import Generic, TypeVar
from junglescout.session import BaseSession

T = TypeVar("T", bound=BaseSession)

class BaseClient(ABC, Generic[T]):
    def __init__(
        self,
        api_key_name: str,
        api_key: str,
        api_type: ApiType = ApiType.JS,
        marketplace: Optional[Marketplace] = None,
    ):
        self.session: T = self.create_session()
        self.session.login(api_key_name=api_key_name, api_key=api_key, api_type=api_type)

    @abstractmethod
    def create_session(self) -> T:
        pass
```

Then something like this for the subclass:

```python
class AsyncClient(BaseClient[AsyncSession]):
    def __init__(
        self,
        api_key_name: str,
        api_key: str,
        api_type: ApiType = ApiType.JS,
        marketplace: Optional[Marketplace] = None,
    ):
        super().__init__(api_key_name, api_key, api_type, marketplace)

    def create_session(self) -> AsyncSession:
        headers = self._build_headers()
        session = AsyncSession(headers=headers)
        return session
```

Reply:
True, I prefer you implementation